lib/cpu.h

```diff
@@ -18,6 +18,13 @@
 # include <metal/processor/@PROJECT_PROCESSOR@/cpu.h>
 #else
 # include <metal/processor/generic/cpu.h>
+# include <metal/sleep.h>
+#endif
+
+#if defined(HAVE_PROCESSOR_CPU_H)
+#define metal_yield() metal_cpu_yield()
+#else
+#define metal_yield() metal_sleep_usec(1000)
```

Comment:
Seems dangerous to use HAVE_PROCESSOR_CPU_H. just define metal_yield in sys.h files an d redirect it to  metal_cpu_yield except for Nuttx ( I suppose it is for NuttX).

here is a proposed patch:

```
diff --git a/lib/system/freertos/sys.h b/lib/system/freertos/sys.h
index d0d64b550e6d..2d671c9fa937 100644
--- a/lib/system/freertos/sys.h
+++ b/lib/system/freertos/sys.h
@@ -18,6 +18,7 @@
 #define __METAL_FREERTOS_SYS__H__
 
 #include <metal/errno.h>
+#include <metal/cpu.h>
 
 #ifdef XLNX_PLATFORM
 #include <metal/system/freertos/xlnx/sys.h>
@@ -40,6 +41,8 @@ struct metal_state {
 	struct metal_common_state common;
 };
 
+#define metal_yield() metal_cpu_yield()
+
 #ifdef METAL_INTERNAL
 
 /**
diff --git a/lib/system/generic/template/sys.h b/lib/system/generic/template/sys.h
index d5faf83c8041..242fa174169a 100644
--- a/lib/system/generic/template/sys.h
+++ b/lib/system/generic/template/sys.h
@@ -16,10 +16,14 @@
 #ifndef __METAL_GENERIC_TEMPLATE_SYS__H__
 #define __METAL_GENERIC_TEMPLATE_SYS__H__
 
+#include <metal/cpu.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+#define metal_yield() metal_cpu_yield()
+
 #ifdef METAL_INTERNAL
 
 void sys_irq_enable(unsigned int vector);
diff --git a/lib/system/generic/xlnx/sys.h b/lib/system/generic/xlnx/sys.h
index 0b352de6fa85..3f4db9691dac 100644
--- a/lib/system/generic/xlnx/sys.h
+++ b/lib/system/generic/xlnx/sys.h
@@ -23,6 +23,7 @@
 #include <metal/sys.h>
 #include <stdint.h>
 #include <metal/compiler.h>
+#include <metal/cpu.h>
 #include <metal/utilities.h>
 
 #include <xil_cache.h>
@@ -40,6 +41,8 @@
 extern "C" {
 #endif
 
+#define metal_yield() metal_cpu_yield()
+
 #ifndef __MICROBLAZE__
 #define XLNX_MAXIRQS XSCUGIC_MAX_NUM_INTR_INPUTS
 #endif
diff --git a/lib/system/linux/sys.h b/lib/system/linux/sys.h
index 41d1dcdc6d80..e0bf9213aa8b 100644
--- a/lib/system/linux/sys.h
+++ b/lib/system/linux/sys.h
@@ -19,6 +19,7 @@
 #include <fcntl.h>
 #include <libgen.h>
 #include <limits.h>
+#include <metal/cpu.h>
 #include <metal/errno.h>
 #include <pthread.h>
 #include <stdio.h>
@@ -46,6 +47,8 @@ extern "C" {
 #define METAL_INVALID_VADDR     NULL
 #define MAX_PAGE_SIZES		32
 
+#define metal_yield() metal_cpu_yield()
+
 struct metal_device;
 
 /** Structure of shared page or hugepage sized data. */
diff --git a/lib/system/nuttx/sys.h b/lib/system/nuttx/sys.h
index de19568b0b38..bb598dd2b4a4 100644
--- a/lib/system/nuttx/sys.h
+++ b/lib/system/nuttx/sys.h
@@ -20,6 +20,8 @@
 extern "C" {
 #endif
 
+#define metal_yield() metal_sleep_usec(1000)
+
 #define METAL_INIT_DEFAULTS				\
 {							\
 	.log_handler	= (metal_log_handler)syslog,	\
-
```

Reply:
This does look better and has been applied to this PR.